File: training_data/reflection-tuning/day23_part1_2015.py
def execute_program(instructions):
    registers = {'a': 0, 'b': 0}
    index = 0

    def hlf(r):
        registers[r] //= 2
        return index + 1

    def tpl(r):
        registers[r] *= 3
        return index + 1

    def inc(r):
        registers[r] += 1
        return index + 1

    def jmp(offset):
        return index + int(offset)

    def jie(r, offset):
        return index + int(offset) if registers[r] % 2 == 0 else index + 1

    def jio(r, offset):
        return index + int(offset) if registers[r] == 1 else index + 1

    while 0 <= index < len(instructions):
        instruction = instructions[index].split()
        op = instruction[0]
        if op == 'hlf':
            index = hlf(instruction[1])
        elif op == 'tpl':
            index = tpl(instruction[1])
        elif op == 'inc':
            index = inc(instruction[1])
        elif op == 'jmp':
            index = jmp(instruction[1])
        elif op == 'jie':
            index = jie(instruction[1][0], instruction[2])
        elif op == 'jio':
            index = jio(instruction[1][0], instruction[2])

    return registers['b']

# Read instructions from file
with open('input.txt', 'r') as file:
    instructions = file.read().splitlines()

# Execute program and print result
result = execute_program(instructions)
print(result)

File: fortran90/day23_part1_2015.f90

program instructions
    implicit none
    integer, parameter :: max_lines = 1000
    character(len=3) :: ops(max_lines)
    character(len=1) :: regs(max_lines)
    integer :: offsets(max_lines)
    integer :: a, b, i, n, offset
    character(len=20) :: line
    
    open(unit=10, file="input.txt", status="old")
    n = 0
    do
        read(10, '(A)', end=100) line
        n = n + 1
        if (line(1:3) == 'hlf' .or. line(1:3) == 'tpl' .or. line(1:3) == 'inc') then
            ops(n) = line(1:3)
            regs(n) = line(5:5)
        else if (line(1:3) == 'jmp') then
            ops(n) = line(1:3)
            read(line(5:), *) offsets(n)
        else if (line(1:3) == 'jie' .or. line(1:3) == 'jio') then
            ops(n) = line(1:3)
            regs(n) = line(5:5)
            read(line(8:), *) offsets(n)
        end if
    end do
    
100 close(10)
    
    a = 0
    b = 0
    i = 1
    
    do while (i <= n)
        if (ops(i) == 'hlf') then
            if (regs(i) == 'a') a = a / 2
            if (regs(i) == 'b') b = b / 2
        else if (ops(i) == 'tpl') then
            if (regs(i) == 'a') a = a * 3
            if (regs(i) == 'b') b = b * 3
        else if (ops(i) == 'inc') then
            if (regs(i) == 'a') a = a + 1
            if (regs(i) == 'b') b = b + 1
        else if (ops(i) == 'jmp') then
            i = i + offsets(i) -1
        else if (ops(i) == 'jie') then
            if (regs(i) == 'a' .and. mod(a, 2) == 0) i = i + offsets(i) - 1
            if (regs(i) == 'b' .and. mod(b, 2) == 0) i = i + offsets(i) - 1
        else if (ops(i) == 'jio') then
            if (regs(i) == 'a' .and. a == 1) i = i + offsets(i) - 1
            if (regs(i) == 'b' .and. b == 1) i = i + offsets(i) - 1
        end if
        i = i + 1
    end do
    
    print *, b
    
end program instructions
